File: 1006-vowel-spellchecker/vowel-spellchecker.java
class Solution {
    public String[] spellchecker(String[] wordlist, String[] queries) {
        Set<String> exact = new HashSet<>(Arrays.asList(wordlist));
        Map<String, String> caseMap = new HashMap<>();
        Map<String, String> vMap = new HashMap<>();

        for(String word: wordlist){
            String lowerC = word.toLowerCase();
            caseMap.putIfAbsent(lowerC, word);
            vMap.putIfAbsent(asTreic(lowerC), word);
        }

        String[] result = new String[queries.length];
        for(int i=0; i<result.length; i++){
            String queri = queries[i];
            if(exact.contains(queri)) result[i] = queri;
            else{
                String lowerQueri = queri.toLowerCase();
                if(caseMap.containsKey(lowerQueri)) result[i] = caseMap.get(lowerQueri);
                else{
                    String astreic = asTreic(lowerQueri);
                    result[i] = vMap.getOrDefault(astreic, "");
                }
            }
        }
        return result;
    }

    public String asTreic(String s){
        StringBuilder sb = new StringBuilder();
        for(char ch: s.toCharArray()){
            if("aeiou".indexOf(ch) != -1){
                sb.append("*");
            }
            else sb.append(ch);
        }
        return sb.toString();
    }
}

File: 1006-vowel-spellchecker/vowel-spellchecker.cpp


class Solution {
public:
    vector<string> spellchecker(vector<string>& wordlist, vector<string>& queries) {
        unordered_set<string> exact(wordlist.begin(), wordlist.end());
        unordered_map<string, string> caseMap;
        unordered_map<string, string> vMap;

        for (auto &word : wordlist) {
            string lowerC = toLower(word);
            if (caseMap.find(lowerC) == caseMap.end())
                caseMap[lowerC] = word;
            string ast = asTreic(lowerC);
            if (vMap.find(ast) == vMap.end())
                vMap[ast] = word;
        }

        vector<string> result;
        for (auto &queri : queries) {
            if (exact.count(queri)) {
                result.push_back(queri);
            } else {
                string lowerQueri = toLower(queri);
                if (caseMap.count(lowerQueri)) {
                    result.push_back(caseMap[lowerQueri]);
                } else {
                    string ast = asTreic(lowerQueri);
                    if (vMap.count(ast)) result.push_back(vMap[ast]);
                    else result.push_back("");
                }
            }
        }
        return result;
    }

private:
    string toLower(const string &s) {
        string res = s;
        for (auto &ch : res) ch = tolower(ch);
        return res;
    }

    string asTreic(const string &s) {
        string res;
        for (char ch : s) {
            if (isVowel(ch)) res.push_back('*');
            else res.push_back(ch);
        }
        return res;
    }

    bool isVowel(char ch) {
        return ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u';
    }
};
